WebContent/zip-ext.js

```diff
@@ -208,11 +208,165 @@
 	FileWriter.prototype = new Writer();
 	FileWriter.prototype.constructor = FileWriter;
 
+	function DBWriter(contentType) {
+		var db, tempDB, that = this, blobs, dbName = "zipjs", instance;
+
+		function init(callback, onerror) {
+			var request = indexedDB.open(dbName, 5);
+			request.onerror = onerror;
+			request.onupgradeneeded = function (event) {
+				db = event.target.result;
+				db.createObjectStore("instances", { autoIncrement: true });
+			};
+			request.onsuccess = function (event) {
+				db = event.target.result;
+				addInstance(callback, onerror);
+			};
+			blobs = [];
+		}
+
+		function addInstance(callback, onerror) {
+			var t = db.transaction(["instances"], "readwrite"),
+				request = t.objectStore("instances").put(Date.now());
+			request.onerror = onerror;
+			request.onsuccess = function (event) {
+				instance = request.result;
+				cleanUpOldInstances();
+				window.addEventListener("storage", pongDB);
+				window.addEventListener("unload", close);
+				addInstanceDB(callback, onerror);
+			};
+		}
+
+		function addInstanceDB(callback, onerror) {
+			var request;
+			try {
+				request = indexedDB.open(dbName + "_" + instance, {temporary: true});
+			} catch (e) {
+				request = indexedDB.open(dbName + "_" + instance);
+			}
+			request.onerror = onerror;
+			request.onupgradeneeded = function (event) {
+				tempDB = event.target.result;
+				tempDB.createObjectStore("files", { autoIncrement: true });
+			};
+			request.onsuccess = function (event) {
+				callback();
+			};
+		}
+
+		function close() {
+			blobs = [];
+			indexedDB.deleteDatabase(dbName + "_" + instance);
+			window.removeEventListener("storage", pongDB);
+			window.removeEventListener("unload", pongDB);
```

Comment:
And also remove the event listeners before deleting the database. That will clear any doubts about re-entrance (i.e. another tab simultaneously triggers the storage event between line 260 and 261, which would cause the database to be restored again). I'm not entirely sure whether this happens in practice, but you'd better be safe than sorry.